### PR TITLE
Poetry packaging

### PR DESCRIPTION
**Pull request by bwheelz36**
In this PR, I have recast the packaging mechanism to user [poetry](https://python-poetry.org/). This brings us line with latest python recommendations around packaging using a `pyproject.toml` file. Note that `setup.py` which we were using is considered deprecated. I have also made a few other changes:

- removed some other CI config files which aren't being used anymore
- bumped our minimum supported python version to 3.8, both in our packaging and in our CI. Note that 3.7 is now deprecated.
- bumped the minimum support scikit-learn version from >0.18 to >1.0

This PR paves the way for another change I would like to make: add a CI job for 'release' as described in #426. This would mean that when we release a package through GH releases, a version will also be automatically uploaded to PyPi.

### CODE REVIEW

**Comment by bwheelz36** [comment]
Update: also had to remove python3.8 from supported packages. Poetry is quite strict about compatibility, and there is no version of numpy which supports both 3.8 and 3.12. 3.8 is end of life in October this year, so I think this is ok....  (noting that previous PyPi packages will still be available and support older python versions).

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+description = "Bayesian Optimization package"
+authors = ["Fernando Nogueira"]
+license = "MIT"
+readme = "README.md"
+packages = [{include = "bayes_opt"}]
+
+[tool.poetry.dependencies]
+python = "^3.9"
+scikit-learn = ">=1.0.0"
+# numpy = ">=1.9.0"
```
Comment (till-m) [inline pyproject.toml:13]: Should this be commented out?

**Comment by till-m** [comment]
At the risk of asking a potentially stupid question -- is it possible to host docs for multiple releases in parallel such that one can view the docs of a specific version instead of always viewing the most current one?

**Comment by bwheelz36** [comment]
Hey @till-m 

- **numpy depedency** good catch, it was an accident to comment it out . Since numpy is a dependency of scipy we can get away with this, but it's bad practice to rely on another package to handle a dependency we use explicitly - I will fix.
- **multiple doc versions** - this must be possible because readthedocs supports it. however, tbh I don't have the bandwidth to look at implementing it. For now, previous versions of the docs would always be available via the commit history on the `gh-pages` branch - which is clunky but functional I guess.
- **notebook changes** - these are trivial, I was just re-running to make sure they still ran, no actual changes were made. I probably shouldn't have committed those changes sorry.
- Yes, I won't merge until you give the ok :-)

File: bayes_opt/__init__.py [hunk 2]
```diff
@@ -6,6 +6,12 @@
 from .constraint import ConstraintModel
 from .util import UtilityFunction
 
+import importlib.metadata
+# __version__ = importlib.metadata.version(__name__)
```
[the remark this answers is not in the document]
Comment (bwheelz36) [reply bayes_opt/__init__.py]: Hey @till-m,
Slightly annoyingly, the src folder names is `bayes_opt`, however the package name (as defined on pypi and in `pyproject.toml` is `bayesian-optimization` which was causing this line to crash.

I've fixed this now by hardcoding the name of the package into `__init__`.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+[tool.poetry]
+name = "bayesianoptimization"
```
[the remark this answers is not in the document]
Comment (bwheelz36) [reply pyproject.toml]: yes!

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+[tool.poetry.group.nbtools]  # for running/converting notebooks
+optional = true
+[tool.poetry.group.nbtools.dependencies]
+nbformat = "^5.9.2"
+nbconvert = "^7.14.2"
+jupyter = "^1.0.0"
+matplotlib = "^3.0"
+
+
```
[the remark this answers is not in the document]
Comment (bwheelz36) [reply pyproject.toml:36]: I've added the doc tools into the `dev` group - we could also make it a separate group if you prefer.

**Review by till-m — APPROVED** [review]
Didn't catch anything else. LGTM! :)

**Pull request merged**